### ISSUE TITLE
ActiveRecord Regression 5.2 - update alters query default scope incorrectly

### ISSUE THREAD

**Issue by yosiat** [rails/rails#33470]
While trying to upgrade rails app from 5.1 to rails 5.2 I found some weird behaviour change.

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "5.2.0"
  gem "sqlite3"
  gem "byebug"
end

require "active_record"
require "logger"
require "byebug"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.integer :another_post_id
    t.string :title
  end
end

class Post < ActiveRecord::Base
  belongs_to :another_post, class_name: "Post", foreign_key: "another_post_id"

  before_update :log_data


  def update_all_posts(except_ids)
    Post.where.not(id: except_ids).update(title: "hello")
  end

  def log_data
    if id == 2
      # as a result of update_all_posts, all queries for Post, have default query scope of "id != 1"
      # so we can't find another_post (since it's id 1)
      puts "Me: #{id}, Another Post (#{another_post_id}): #{another_post.inspect}"
      # Post.count is 1, but actually it's 2
      puts "Post.count: #{Post.count}"
    end
  end
end

ActiveRecord::Base.logger = Logger.new(STDOUT)

post1 = nil
post2 = nil

post1 = Post.create!
post2 = Post.create!
post2.another_post_id = post1.id
post2.save

post2.update_all_posts([post1.id])
```
The output for the above code in Rails 5.2:
```
Me: 2, Another Post (1): #<Post id: 1, another_post_id: nil, title: nil>
Post.count: 2

Me: 2, Another Post (1): nil
Post.count: 1
```

I noticed that `Post.where.not` changes the scope of next queries in the callback.
In the above code example, I call `update_all_posts` to update all posts that doesn't match set of ids, as a reuslt `log_data` callback is called and prints wrong informaiton - 

- `another_post` is nil
- `Post.count` is 1, while there are two

While debugging, I see that `Post.unscoped.count` returns 2, and I see that queries for `Post.count` and `another_post` have `WHERE id != 1`.

### Expected behavior

Expected behavior can be observed while running in rails 5.1.6:

```
Me: 2, Another Post (1): #<Post id: 1, another_post_id: nil, title: nil>
Post.count: 2

Me: 2, Another Post (1): #<Post id: 1, another_post_id: nil, title: nil>
Post.count: 2
```

### Temporary fix

A temporary fix I found for this issue, is changing:
```ruby
Post.where.not(id: except_ids).update(title: "hello")
```

To:
```ruby
Post.where.not(id: except_ids).each { |p| p.update(title: "hello") }
```

### System configuration
**Rails version**: 5.2.0

**Ruby version**: 2.5.1p57

**Comment by yosiat** [comment]
Amazing! Thanks @kamipo for the fix!

Is there any plan to release in 5.2.1, if so do we have an estimation for rails 5.2.1 release date?

**Comment by kaspth** [comment]
@yosiat you might want to keep an eye on the weblog: https://weblog.rubyonrails.org/2018/7/30/Rails-5-2-1-rc1-has-been-released/ 😉

**Comment by kaspth** [comment]
Though @kamipos commit isn't on 5-2-stable, yet. So it won't make it into 5.2.1.

**Comment by yosiat** [comment]
If it's is possible it will be amazing for it to be in rails 5.2.1, otherwise we will need to skip version in rails and wait for 6, since this is blocker for us

**Comment by sikachu** [comment]
Yes, I think we will backport this to `5-2-stable` and `5-2-1`, so it should be out with the next RC/5.2.1 final release.

**Comment by kamipo** [comment]
Yeah, I've backported to [5-2-stable](https://github.com/rails/rails/commit/dd1133699f7266e9659c3624de1c77fcd43a4691) and [5-2-1](https://github.com/rails/rails/commit/0b29a421ec376a413ff8902b692d8a29afaa4b51).

**Comment by yosiat** [comment]
Thanks @kamipo !

**Comment by rafaelfranca** [comment]
This commit cause a regression in our codebase so it is not safe to be included in the 5.2.1 release without a RC so I'm reverting.

This commit beaks when the `update` method is called with two arguments, the id and the attributes in a relation, or an association. This problem is also present on 5-2-stable and we should either revert or support that feature.

**Comment by kamipo** [comment]
I can support that feature, but the `update` with two arguments is designed feature?
At least that usage is not documented.

https://github.com/rails/rails/blob/a6e86dd2a86a89b7775e27a2f0e728cb6afd1513/activerecord/lib/active_record/persistence.rb#L84-L110

**Comment by rafaelfranca** [comment]
We could say that `Person.update(people.keys, people.values)` make we think it is a designed feature.  Given that `create` in the class and the relation behave in the same way `update` in the class and relation should behave in the same way.

I'm fine if we remove that support but we should not remove in a stable branch and neither without deprecation on master. This also means that `update` in the relation should be documented with a different API that could be confusing given people will not know which one they want to use. I prefer to keep the behavior as it is today with both the class and the relation methods having the same behavior

**Comment by kamipo** [comment]
I see... so I'll working on that to make the `update` with two arguments in a relation promote to public API.

**Comment by yosiat** [comment]
@rafaelfranca @kamipo since it's reverted from rails 5.2.1, do we have expected release version? or it will be fixed in the next major version 6.0?

Currently, we are blocked on upgrading from rails 5.1 to 5.2 and I want to better understand what's the status and how can I help here

**Comment by kamipo** [comment]
I'll be going to address Shopify's CI failure, so probably this could be fixed in next 5.2 release.

**Comment by yosiat** [comment]
Thanks @kamipo can't wait for it!
